FinniversKit/Sources/Fullscreen/FrontPage/PromotionFrontPageView.swift

```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    public func reloadData() {
+        marketsView.reloadData()
+        setupFrames()
+        reloadAds()
+    }
+
+    public func reloadMarkets() {
+        marketsView.reloadData()
+        setupFrames()
+        adRecommendationsGridView.reloadData()
+    }
+
+    public func reloadAds() {
+        adsRetryView.state = .hidden
+        adRecommendationsGridView.reloadData()
+    }
+
+    public func updateAd(at index: Int, isFavorite: Bool) {
+        adRecommendationsGridView.updateItem(at: index, isFavorite: isFavorite)
+    }
+
+    public func showAdsRetryButton() {
+        adRecommendationsGridView.endRefreshing()
+        adsRetryView.state = .labelAndButton
+    }
+
+    public func scrollToTop() {
+        adRecommendationsGridView.scrollToTop()
+    }
+
+    // MARK: - Setup
+
+    private func setup() {
+        backgroundColor = .bgPrimary
+
+        addSubview(adRecommendationsGridView)
+        adRecommendationsGridView.fillInSuperview()
+        adRecommendationsGridView.collectionView.addSubview(adsRetryView)
+        adRecommendationsGridView.collectionView.addSubview(promoView)
+        adRecommendationsGridView.collectionView.addSubview(marketsView)
+
+        adRecommendationsGridView.collectionView.bringSubviewToFront(promoView)
+        adRecommendationsGridView.collectionView.bringSubviewToFront(marketsView)
+
+        NSLayoutConstraint.activate([
+            promoViewTopConstraint,
+            promoView.leadingAnchor.constraint(equalTo: adRecommendationsGridView.leadingAnchor),
+            promoView.trailingAnchor.constraint(equalTo: adRecommendationsGridView.trailingAnchor),
+            promoView.heightAnchor.constraint(equalToConstant: promoViewHeight),
+
+            marketsView.topAnchor.constraint(equalTo: promoView.bottomAnchor),
+            marketsView.leadingAnchor.constraint(equalTo: adRecommendationsGridView.leadingAnchor),
+            marketsView.trailingAnchor.constraint(equalTo: adRecommendationsGridView.trailingAnchor),
+            marketsViewHeightConstraint,
+        ])
+
+        setupFrames()
+    }
+
+    private func setupFrames() {
+        let marketsViewHeight = marketsView.calculateSize(constrainedTo: bounds.size.width).height
+        marketsViewHeightConstraint.constant = marketsViewHeight
+        adsRetryView.frame.origin = .zero
+        adsRetryView.frame.size = CGSize(width: bounds.width, height: 200)
+        boundsForCurrentSubviewSetup = bounds
+        adRecommendationsGridView.invalidateLayout()
+        adRecommendationsGridView.collectionView.contentInset.top = marketsViewHeight + promoViewHeight
+    }
+}
+
+// MARK: - LoadingRetryViewDelegate
+
+extension PromotionFrontPageView: LoadingRetryViewDelegate {
+    func loadingRetryViewDidSelectButton(_ view: LoadingRetryView) {
+        adsRetryView.state = .loading
+        delegate?.promotionFrontPageViewDidSelectRetryButton(self)
+    }
+}
+
+// MARK: - AdRecommendationsGridViewDelegate proxy
```

Comment:
proxy? 💭

Reply:
Yes, `PromotionFrontPageView` is implementing `AdRecommendationsGridViewDelegate`, but forwarding all calls to it. The only method I need to act on is this one to get to know the scroll offset:

```swift
public func adRecommendationsGridView(
    _ adRecommendationsGridView: AdRecommendationsGridView,
    didScrollInScrollView scrollView: UIScrollView
)
```

Reply:
aha, I understand 😄 👍